dns-secondary-operator/src/charm.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm for dns-secondary."""
+
+import logging
+import subprocess  # nosec
+import time
+import typing
+
+import ops
+from charms.dns_transfer.v0 import dns_transfer
+
+import constants
+import events
+import exceptions
+import topology
+from bind import BindService
+
+logger = logging.getLogger(__name__)
+
+STATUS_REQUIRED_INTEGRATION = "Required integration with DNS primary not found"
+
+
+class PeerRelationUnavailableError(exceptions.DnsSecondaryCharmError):
+    """Exception raised when the peer relation is unavailable."""
+
+
+class PeerRelationNetworkUnavailableError(exceptions.DnsSecondaryCharmError):
+    """Exception raised when the peer relation network is unavailable."""
+
+
+class DnsSecondaryCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+        self.bind = BindService()
+        self.topology = topology.TopologyObserver(self, constants.PEER)
```

Comment:
The TopologyObserver is outputting topology_changed events that you may want to listen

Reply:
Good catch!